k2/csrc/rnnt_decode.h

```diff
@@ -0,0 +1,390 @@
+/**
+ * Copyright      2022  Xiaomi Corporation (authors:Daniel Povey, Wei kang)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef K2_CSRC_RNNT_DECODE_H_
+#define K2_CSRC_RNNT_DECODE_H_
+
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+#include "k2/csrc/array.h"
+#include "k2/csrc/array_of_ragged.h"
+#include "k2/csrc/array_ops.h"
+#include "k2/csrc/log.h"
+#include "k2/csrc/macros.h"
+
+namespace k2 {
+namespace rnnt_decoding {
+
+/*
+  The RNN-T decoding implemented here is for what we call "modified" RNN-T, or
+  equivalently, regular RNN-T but with max_sym_per_frame set to 1.  (We can
+  train with the "modified" option set with some probability, in order to ensure
+  that the trained model is compatible with decoding with this setting).
+
+  - contexts are finite symbol left-contexts, of length
+  RnntDecodingConfig::decoder_history_len. conceptuatly they represent a list of
+  `decoder_history_len` symbols; they are represented numerically as, for
+  example in the length-2-history case: symbol_{t-1} +  symbol_{t-2} *
+  vocab_size.
+
+  - frames come from the transcription network, they are derived from
+  sub-sampling of acoustic frames or samples.
+ */
+
+struct RnntDecodingConfig {
+  RnntDecodingConfig(int32_t vocab_size, int32_t decoder_history_len,
+                     double beam, int32_t max_states, int32_t max_contexts)
+      : vocab_size(vocab_size),
+        decoder_history_len(decoder_history_len),
+        beam(beam),
+        max_states(max_states),
+        max_contexts(max_contexts) {
+    num_context_states = pow(vocab_size, decoder_history_len);
+  }
+
+  // vocab_size is the largest-symbol plus one.
+  int32_t vocab_size;
+
+  // decoder_history_len is the number of symbols
+  // of history the decoder takes; will normally
+  // be one or two ("stateless decoder"), this
+  // RNN-T decoding setup does not support
+  // unlimited decoder context such as with LSTMs
+  int32_t decoder_history_len;
+
+  // num_context_states == pow(vocab_size, decoder_history_len).
+  // We need an unique id for each context state, think about that if vocab_size
+  // equals to 10, we need 0 ~ 9 to distinguish each context state when
+  // decoder_history_len is 1, and 0 ~ 99 (10 ^ 2 ids) for decoder_history_len
+  // equals to 2, 0 ~ 999 (10 ^ 3 ids) for decoder_history_len equals to 3.
+  int32_t num_context_states;
```

Comment:
Looks like `num_context_states` is used nowhere.
Shall we delete it?

Reply:
.. could comment it out perhaps.  The concept may be useful, and it may be referred to in other docs.